src/unix/signal.rs

```diff
@@ -402,6 +413,77 @@ pub fn clear_signal(num: c_int) -> SignalResult<()> {
     Ok(())
 }
 
+/// Suspend execution until a signal is pending.
+///
+/// According to the [man page](https://linux.die.net/man/2/sigtimedwait), the caller should block
+/// the signals prior to waiting on them via [`block_signal`](fn.block_signal.html) and should make
+/// sure that no signal handlers are configured for them.
+///
+/// # Arguments
+///
+/// * `signals`: the signals that unblock execution. If none are specified, any pending signal will
+///              unblock execution.
+/// * `timeout_nanos`: optional timeout to wait for a signal to be pending. If `None` is specified,
+///                    execution remains suspended until a signal arrives.
+///
+/// # Examples
+///
+/// ```rust
+/// # extern crate libc;
+/// extern crate vmm_sys_util;
+/// # use std::thread;
+/// # use std::time::Duration;
+/// use vmm_sys_util::signal::{block_signal, wait_for_signals, Killable, SIGRTMIN};
+///
+/// let signal = SIGRTMIN() + 2;
+/// let signals = vec![signal];
+/// block_signal(signal).unwrap();
+/// let killable = thread::spawn(move || {
+///     assert!(wait_for_signals(&signals, None).is_ok());
+/// });
+/// thread::sleep(Duration::from_millis(100));
+/// assert!(killable.kill(signal).is_ok());
+/// killable.join().unwrap();
+///
+/// let signals = vec![signal];
+/// let killable = thread::spawn(move || {
+///     assert!(wait_for_signals(&signals, Some(500_000_000)).is_ok());
+/// });
+/// thread::sleep(Duration::from_millis(600));
+/// killable.join().unwrap();
+/// ```
+pub fn wait_for_signals(signals: &[c_int], timeout_nanos: Option<c_longlong>) -> SignalResult<()> {
+    let mut sigset: sigset_t = create_sigset(signals).map_err(Error::CreateSigset)?;
+    if signals.is_empty() {
+        unsafe { sigfillset(&mut sigset) };
+    }
+
+    if let Some(nanos) = timeout_nanos {
+        let duration = timespec {
+            tv_sec: nanos / 1_000_000_000,
+            tv_nsec: nanos % 1_000_000_000,
+        };
+        let ret = unsafe { sigtimedwait(&sigset, null_mut(), &duration) };
+        if ret < 0 {
+            let e = errno::Error::last();
+            match e.errno() {
+                // `EAGAIN` means no signal became pending in the duration specified.
+                EAGAIN => (),
```

Comment:
Do we need to return the TimedOut condition to caller?

Reply:
`EAGAIN` is not an error condition, it's an indicator that no signal stopped the wait. Any other error (according to the man page, it can be either `EINVAL` or `EINTR`) will be passed to the user 2 lines below.